File: python/sigil_sdk/tracing.py
"""OpenTelemetry trace + metrics runtime wiring."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Callable
from urllib.parse import urlparse

from opentelemetry import metrics, trace
from opentelemetry.exporter.otlp.proto.grpc.metric_exporter import OTLPMetricExporter as OTLPGRPCMetricExporter
from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import OTLPSpanExporter as OTLPGRPCSpanExporter
from opentelemetry.exporter.otlp.proto.http.metric_exporter import OTLPMetricExporter as OTLPHTTPMetricExporter
from opentelemetry.exporter.otlp.proto.http.trace_exporter import OTLPSpanExporter as OTLPHTTPSpanExporter
from opentelemetry.sdk.metrics import MeterProvider
from opentelemetry.sdk.metrics.export import PeriodicExportingMetricReader
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor

from .config import TraceConfig


INSTRUMENTATION_NAME = "github.com/grafana/sigil/sdks/python"


@dataclass(slots=True)
class TraceRuntime:
    """Managed trace + metrics runtime resources."""

    tracer: trace.Tracer
    meter: metrics.Meter
    trace_provider: TracerProvider | None
    meter_provider: MeterProvider | None

    def flush(self) -> None:
        if self.trace_provider is not None:
            self.trace_provider.force_flush()
        if self.meter_provider is not None:
            self.meter_provider.force_flush()

    def shutdown(self) -> None:
        if self.trace_provider is not None:
            self.trace_provider.shutdown()
        if self.meter_provider is not None:
            self.meter_provider.shutdown()


def create_trace_runtime(config: TraceConfig, on_error: Callable[[str, Exception], None] | None = None) -> TraceRuntime:
    """Builds a tracer/meter runtime pair for the configured OTLP transport."""

    try:
        trace_exporter = _new_trace_exporter(config)
        trace_provider = TracerProvider()
        trace_provider.add_span_processor(BatchSpanProcessor(trace_exporter))

        metric_exporter = _new_metric_exporter(config)
        metric_reader = PeriodicExportingMetricReader(metric_exporter)
        meter_provider = MeterProvider(metric_readers=[metric_reader])

        return TraceRuntime(
            tracer=trace_provider.get_tracer(INSTRUMENTATION_NAME),
            meter=meter_provider.get_meter(INSTRUMENTATION_NAME),
            trace_provider=trace_provider,
            meter_provider=meter_provider,
        )
    except Exception as exc:  # noqa: BLE001
        if on_error is not None:
            on_error("sigil telemetry exporter init failed", exc)
        return TraceRuntime(
            tracer=trace.get_tracer(INSTRUMENTATION_NAME),
            meter=metrics.get_meter(INSTRUMENTATION_NAME),
            trace_provider=None,
            meter_provider=None,
        )


def _new_trace_exporter(config: TraceConfig):
    protocol = config.protocol.strip().lower()
    if protocol == "grpc":
        endpoint, implicit_insecure = _parse_endpoint(config.endpoint)
        insecure = config.insecure or implicit_insecure
        return OTLPGRPCSpanExporter(
            endpoint=endpoint,
            insecure=insecure,
            headers=dict(config.headers),
            timeout=10,
        )

    endpoint, implicit_insecure = _parse_endpoint(config.endpoint)
    if "://" not in endpoint:
        scheme = "http" if (config.insecure or implicit_insecure) else "https"
        endpoint = f"{scheme}://{endpoint}"
    if urlparse(endpoint).path in ("", "/"):
        endpoint = endpoint.rstrip("/") + "/v1/traces"

    return OTLPHTTPSpanExporter(
        endpoint=endpoint,
        headers=dict(config.headers),
        timeout=10,
    )


def _new_metric_exporter(config: TraceConfig):
    protocol = config.protocol.strip().lower()
    if protocol == "grpc":
        endpoint, implicit_insecure = _parse_endpoint(config.endpoint)
        insecure = config.insecure or implicit_insecure
        return OTLPGRPCMetricExporter(
            endpoint=endpoint,
            insecure=insecure,
            headers=dict(config.headers),
            timeout=10,
        )

    endpoint, implicit_insecure = _parse_endpoint(config.endpoint)
    if "://" not in endpoint:
        scheme = "http" if (config.insecure or implicit_insecure) else "https"
        endpoint = f"{scheme}://{endpoint}"

    parsed = urlparse(endpoint)
    if parsed.path in ("", "/", "/v1/traces"):
        endpoint = endpoint.rstrip("/")
        endpoint += "/v1/metrics"

    return OTLPHTTPMetricExporter(
        endpoint=endpoint,
        headers=dict(config.headers),
        timeout=10,
    )


def _parse_endpoint(endpoint: str) -> tuple[str, bool]:
    trimmed = endpoint.strip()
    if not trimmed:
        raise ValueError("trace endpoint is required")

    if "://" not in trimmed:
        return trimmed, False

    parsed = urlparse(trimmed)
    if parsed.netloc == "":
        raise ValueError("trace endpoint host is required")

    if parsed.scheme == "grpc":
        return parsed.netloc, False

    return trimmed, parsed.scheme == "http"

File: js/src/tracing.ts
import * as grpc from '@grpc/grpc-js';
import { metrics, trace, type Meter, type Tracer } from '@opentelemetry/api';
import { OTLPMetricExporter as OTLPMetricExporterGRPC } from '@opentelemetry/exporter-metrics-otlp-grpc';
import { OTLPMetricExporter as OTLPMetricExporterHTTP } from '@opentelemetry/exporter-metrics-otlp-http';
import { OTLPTraceExporter as OTLPTraceExporterGRPC } from '@opentelemetry/exporter-trace-otlp-grpc';
import { OTLPTraceExporter as OTLPTraceExporterHTTP } from '@opentelemetry/exporter-trace-otlp-http';
import { MeterProvider, PeriodicExportingMetricReader } from '@opentelemetry/sdk-metrics';
import { BatchSpanProcessor, BasicTracerProvider, type SpanExporter } from '@opentelemetry/sdk-trace-base';
import type { TraceConfig } from './types.js';

const instrumentationName = 'github.com/grafana/sigil/sdks/js';

export interface TraceRuntime {
  tracer: Tracer;
  meter: Meter;
  flush(): Promise<void>;
  shutdown(): Promise<void>;
}

export function createTraceRuntime(
  config: TraceConfig,
  onError?: (message: string, error: unknown) => void
): TraceRuntime {
  try {
    const traceExporter = createTraceExporter(config);
    const traceProvider = new BasicTracerProvider({
      spanProcessors: [
        new BatchSpanProcessor(traceExporter, {
          maxQueueSize: 2_048,
          maxExportBatchSize: 512,
          scheduledDelayMillis: 1_000,
          exportTimeoutMillis: 1_000,
        }),
      ],
    });

    const metricExporter = createMetricExporter(config);
    const metricReader = new PeriodicExportingMetricReader({
      exporter: metricExporter,
      exportIntervalMillis: 1_000,
      exportTimeoutMillis: 1_000,
    });
    const meterProvider = new MeterProvider({
      readers: [metricReader],
    });

    return {
      tracer: traceProvider.getTracer(instrumentationName),
      meter: meterProvider.getMeter(instrumentationName),
      async flush() {
        await traceProvider.forceFlush();
        await meterProvider.forceFlush();
      },
      async shutdown() {
        await traceProvider.shutdown();
        await meterProvider.shutdown();
      },
    };
  } catch (error) {
    onError?.('sigil telemetry exporter init failed', error);
    return {
      tracer: trace.getTracer(instrumentationName),
      meter: metrics.getMeter(instrumentationName),
      async flush() {},
      async shutdown() {},
    };
  }
}

function createTraceExporter(config: TraceConfig): SpanExporter {
  switch (config.protocol) {
    case 'grpc': {
      const endpoint = parseEndpoint(config.endpoint);
      const url = normalizeGRPCEndpoint(endpoint, config.insecure);
      const metadata = toGRPCMetadata(config.headers);
      const insecure = config.insecure || endpoint.insecure;

      return new OTLPTraceExporterGRPC({
        url,
        metadata,
        credentials: insecure ? grpc.credentials.createInsecure() : grpc.credentials.createSsl(),
        timeoutMillis: 1_000,
      });
    }
    case 'http':
    default:
      return new OTLPTraceExporterHTTP({
        url: normalizeHTTPTraceEndpoint(parseEndpoint(config.endpoint), config.insecure),
        headers: config.headers ? { ...config.headers } : undefined,
        timeoutMillis: 1_000,
      });
  }
}

function createMetricExporter(config: TraceConfig) {
  switch (config.protocol) {
    case 'grpc': {
      const endpoint = parseEndpoint(config.endpoint);
      const url = normalizeGRPCEndpoint(endpoint, config.insecure);
      const metadata = toGRPCMetadata(config.headers);
      const insecure = config.insecure || endpoint.insecure;
      return new OTLPMetricExporterGRPC({
        url,
        metadata,
        credentials: insecure ? grpc.credentials.createInsecure() : grpc.credentials.createSsl(),
        timeoutMillis: 1_000,
      });
    }
    case 'http':
    default:
      return new OTLPMetricExporterHTTP({
        url: normalizeHTTPMetricsEndpoint(parseEndpoint(config.endpoint), config.insecure),
        headers: config.headers ? { ...config.headers } : undefined,
        timeoutMillis: 1_000,
      });
  }
}

function normalizeHTTPTraceEndpoint(endpoint: ParsedEndpoint, insecureConfig: boolean): string {
  const scheme = endpoint.scheme ?? (insecureConfig || endpoint.insecure ? 'http' : 'https');
  const path = endpoint.path.length === 0 || endpoint.path === '/' ? '/v1/traces' : endpoint.path;
  return `${scheme}://${endpoint.host}${path}`;
}

function normalizeHTTPMetricsEndpoint(endpoint: ParsedEndpoint, insecureConfig: boolean): string {
  const scheme = endpoint.scheme ?? (insecureConfig || endpoint.insecure ? 'http' : 'https');
  const trimmedPath = endpoint.path.length > 1 && endpoint.path.endsWith('/')
    ? endpoint.path.slice(0, -1)
    : endpoint.path;
  const path = trimmedPath.length === 0 || trimmedPath === '/'
    ? '/v1/metrics'
    : (trimmedPath === '/v1/traces' || trimmedPath.endsWith('/v1/traces'))
      ? `${trimmedPath.slice(0, -'/v1/traces'.length)}/v1/metrics`
      : trimmedPath;
  return `${scheme}://${endpoint.host}${path}`;
}

function normalizeGRPCEndpoint(endpoint: ParsedEndpoint, insecureConfig: boolean): string {
  const scheme = endpoint.scheme ?? (insecureConfig || endpoint.insecure ? 'http' : 'https');
  return `${scheme}://${endpoint.host}`;
}

function toGRPCMetadata(headers: Record<string, string> | undefined): grpc.Metadata | undefined {
  if (headers === undefined || Object.keys(headers).length === 0) {
    return undefined;
  }

  const metadata = new grpc.Metadata();
  for (const [key, value] of Object.entries(headers)) {
    metadata.set(key, value);
  }
  return metadata;
}

interface ParsedEndpoint {
  host: string;
  path: string;
  scheme?: string;
  insecure: boolean;
}

function parseEndpoint(endpoint: string): ParsedEndpoint {
  const trimmed = endpoint.trim();
  if (trimmed.length === 0) {
    throw new Error('trace endpoint is required');
  }

  if (trimmed.includes('://')) {
    const parsed = new URL(trimmed);
    return {
      host: parsed.host,
      path: parsed.pathname,
      scheme: parsed.protocol.replace(':', ''),
      insecure: parsed.protocol === 'http:',
    };
  }

  const firstSlash = trimmed.indexOf('/');
  if (firstSlash === -1) {
    return {
      host: trimmed,
      path: '',
      insecure: false,
    };
  }

  return {
    host: trimmed.slice(0, firstSlash),
    path: trimmed.slice(firstSlash),
    insecure: false,
  };
}
